Review Java Language Page (#2151)
gitpod-io/website: gitpod/docs/languages/java2.md

https://sdkman.io/---
section: languages
title: Java in Gitpod

---

<script context="module">
  export const prerender = true;
</script>

# Java in Gitpod

Gitpod comes with great support for Java builtin. Still, depending on your particular project you might want to further optimize the experience.

> A great example of a ready-to-code Java dev environment is our [Spring Petclinic Example](https://github.com/gitpod-io/spring-petclinic).

Reviewer: If we are using [Spring Petclinic Example](https://github.com/gitpod-io/spring-petclinic) as example repo for Java dev environment, then i think we should add the implementation of `sdkman` to that repo.

## Ready-to-code Dev Environments for Java

To set up your project, you should run `gp init` in the terminal first. This will create the two configuration files needed to explain Gitpod what your project needs.
A dev environment is based on a Docker image. But don't fear you don't need to understand Docker to get started. In most cases, the default Docker image is enough.

> ### SDKMAN

Author: We should consider making this a more generic section about how to modify base images to ensure you have the correct versions installed, etc.

> ### SDKMAN

Author: **Thought:** How common is it that we expect users to need a version switching utility, as opposed to pinning versions in the docker file / base image? Should we treat base image setup as the first-class?

Reviewer: For Gitpod Enterprise, base image setup won't be helpful. Maybe @julia-leyton have more context to that.

Author: Not sure what you mean, @Siddhant-K-code ? Can you explain a bit more?

Reviewer: Was saying that, there are some special requirements for Java/ JDK/ SDK Versions for particular project. & For Gitpod enterprise, we are also using specific version of Java which requires version switching

Author: Ah, no worries, so I've left SDK MAN in there, explained how you can use it for version switching, but also detailed how you can pin it in the Dockerfile, which should give you the best of both worlds, for now! 🙏

> ### SDKMAN

Reviewer: Yeah, both are Open Source Competitors. Both are good. But, sdkman is more active in `releases`

![image](https://user-images.githubusercontent.com/55068936/172755941-47498412-4f7a-4b5c-a322-93ae43d046a5.png)

> see comparison at [sdkman-cli-vs-jenv](https://java.libhunt.com/compare-sdkman-cli-vs-jenv)

> For Java, it comes readily equipped with [SDKMAN!](https://sdkman.io/), which is a version manager that takes care of managing and installing not only different Java versions but also a different version of other JVM languages such as Scala or Kotlin (both supported in Gitpod, too).

Author: What is "it" in this context?

> It also installs Maven and Gradle in case you don't use the [wrapper version](https://docs.gradle.org/current/userguide/gradle_wrapper.html) that is often recommended.

Author: **Question:** What is it in this context?  SDKMAN? 

**Thought:** We mention the wrapper version, but give no additional context of pro's/con's, etc.

Check out the [documentation of SDKMAN!](https://sdkman.io/usage) to see how to use it or simply type `sdk help` in the terminal.

#### SDKMAN in Docker

Although you can use sdk in your terminal, you should put the tools and versions into your Dockerfile, so you and your team get the very same configuration every time.

> To do that you need to alter the generated `.gitpod.Dockerfile`. Here's an example that will install Java 12 with the J9 VM from "Adopt a JDK".

Author: **Suggestion:** Let's be explicit, and provide copy/pasteable snippets, and explicit steps eg. 

`echo "whatever contents" > .gitpod.Dockerfile`

```dockerfile
FROM gitpod/workspace-full

RUN bash -c ". /home/gitpod/.sdkman/bin/sdkman-init.sh \

>              && sdk install java 12.0.1.j9-adpt"

Author: **Thought:** Why provide just Java 12? Should we provide some variations or different versions here? Or link to the SDK page showing the full list of options.

```

> Note that you always need to run the `sdkman-init.sh` in bash before you can use SDK.

You can add additional tools and versions like this:

```dockerfile
FROM gitpod/workspace-full

RUN bash -c ". /home/gitpod/.sdkman/bin/sdkman-init.sh \
             && sdk install java 12.0.1.j9-adpt \
             && sdk ..."
```

### Prebuilds

> Gitpod provides disposable dev environments, which means you are getting fresh developer environments for every task. So configuring them to be ready-to-code is crucial to get the most out of Gitpod.

Author: > disposable dev environments, which means you are getting fresh developer environments for every task

This statement needs more explanation. 

e.g. Why is this relevant? Why should a user care? Why should a user care in the context of Java?

> So configuring them to be ready-to-code is crucial to get the most out of Gitpod.

Doesn't explain why.

Reviewer: I think, we should discuss the trade-off/ requirement for prebuild or linking to that page. that it requires Gitpod GitHub App to be configured to use & load the previous prebuild. 

reason, i am telling this a trade-off:
 For Enterprise, they will happily configure with Org. (in most cases). But, the scenario is not same for Open Source. OSS Maintainer will not easily agree to configure the Gitpod GitHub App.

Author: So as a "workaround", I introduce the user to the `init` command early on, e.g. 

![image](https://user-images.githubusercontent.com/5528307/174753760-82cfdbeb-aa5f-4663-9695-260d151472d0.png)

But don't mention too much about it, hinting that we'll cover it later. 

![image](https://user-images.githubusercontent.com/5528307/174753834-01238bc7-0797-4ff5-8532-e2a03ce61d2a.png)

Before covering it in a little more detail below in the optimising section. 

![image](https://user-images.githubusercontent.com/5528307/174753938-8687ae31-fa57-4434-88b8-890377e92b15.png)

If you want to add an additional note about the requirements, e.g. GitHub app in some language that you think makes sense for open source, go for it 🙏 Hopefully this will merge soon, we could do a PR / update on top 😁

In the generated `.gitpod.yml` you will find the following section:

```yml

> # List the startup tasks. You can start them in parallel in multiple terminals. See https://www.gitpod.io/docs/config-start-tasks/

Reviewer 2: I would break down long comments in multiple lines, to avoid horizontal scrolling

> ```yml
> # List the startup tasks. You can start them in parallel in multiple terminals. See https://www.gitpod.io/docs/config-start-tasks/

Reviewer: ```suggestion
```yml
# List the startup tasks.
# You can start them in parallel in multiple terminals. 
# See https://www.gitpod.io/docs/config-start-tasks/
```

tasks:
  - init: echo 'init script' # runs during prebuild
    command: echo 'start script'
```

> You can have as many tasks as you which, each will result in an opened terminal when you start a dev environment.

Author: Typo: "Which" => "wish"

Reviewer: ```suggestion
You can have as many tasks as your wish, each task will result in an opened terminal when you start a dev environment.
```

Each task supports multiple phases, most importantly `before`, `init`, `command`. To be ready to code a dev environment should not only be in sync with the Git repo but also needs to have the currently checked out state built.
For instance, in the [Spring Petclinic demo](https://github.com/gitpod-io/spring-petclinic) the command section looks like this:

```yml
# startup tasks
tasks:
  - init: ./mvnw package -DskipTests
    command: java -jar target/*.jar
```

Since we have installed the [Gitpod app](https://github.com/apps/gitpod-io) on that GitHub repository, Gitpod will prebuild any branch as soon as it starts. During a prebuild it will

1.  start a container based on the Docker image,
2.  clone the repository and check out the respective branch,
3.  run the `before` and `init` parts of every task,
4.  capture the result and store it

Once you or your teammates start a dev environment, you will get the prebuild state. The log output from `init` is still presented in the terminal but it will have two additional lines, e.g.:

```sh
🍌 This task ran as part of a workspace prebuild.
🎉 You just saved 4 minutes of watching your code build.
```

> ## IDE features

Author: IDE / Editors

### Debugging

Here is a quick clip on how to automatically configure debugging for Java!

![Java debugging example](../../../static/images/docs/JavaDebug.gif)

So, basically in this video we:

1. First, open the Java file that we want to debug
2. Then, go to the debug menu and select "Add Configuration..."
3. Next, in the dropdown choose "Java: Launch Program in Current File"
4. Finally, start debugging your Java program!

You can also create the Java debug configuration file manually

> To start debugging your Java application in Gitpod, please create a new directory called `.theia/`, and inside add a file called `launch.json`, finally add the following to it:

Author: Remove theia references.

Author: We should definitely bias to IntelliJ, however still worth mentioning some VS Code aspects as we do have users / customers who are using VS Code for Java. Style wise, would be good to re-use the cloud provider toggle, to keep things clean when there are different options per IDE or editor. 

- Issue for toggle with IDE's: https://github.com/gitpod-io/website/issues/2259

![image](https://user-images.githubusercontent.com/5528307/174258838-3094d832-2dbd-424e-a087-2d926cd316a9.png)

| ------------------------------------------------------------------------ | ----------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- |
| [Vaadin CRM](https://github.com/vaadin-learning-center/crm-tutorial)     | Full-stack Vaadin and Spring Boot         | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/vaadin-learning-center/crm-tutorial) |
| [Vaadin starter](https://github.com/vaadin/skeleton-starter-flow-spring) | An empty starter for a new Vaadin project | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/vaadin/skeleton-starter-flow-spring) |

</div>

### Configuring Gitpod for a Vaadin project

Start by downloading a [**Vaadin and Spring Boot** project starter](https://vaadin.com/start) if you don't have a Vaadin project from before.

Next, add a [`.gitpod.yml`](/docs/config-gitpod-file) file with the follwing content to the root of the project:

> ```YAML

Reviewer: ```suggestion
```yml
```

Consistent use of code block type.

tasks:
  - command: mvn spring-boot:run
ports:
  - port: 8080
    onOpen: open-preview
```

This will start the development server and open up the application in the preview browser window.

You are now ready to push your code to GitHub.

### Enable prebuilds for a faster startup

> The first time you start a Vaadin application, it downloads both Maven and npm dependencies, which can take some time. You can speed up the GitPod startup by enabling [prebuild](/docs/prebuilds) for the repository.

Reviewer: ```suggestion
The first time you start a Vaadin application, it downloads both Maven and npm dependencies, which can take some time. You can speed up the Gitpod startup by enabling [prebuild](/docs/prebuilds) for the repository.
```

Update your [`.gitpod.yml`](/docs/config-gitpod-file) file with the following content:

> ```YAML

Reviewer: ```suggestion
```yml
```

tasks:
  - prebuild: mvn install spring-boot:start spring-boot:stop -DskipTests
    command: mvn spring-boot:run
ports:
  - port: 8080
    onOpen: open-preview
github:
  prebuilds:
    master: true
```

Finally, enable prebuilds on GitHub [as instructed in the prebuild documentation](/docs/prebuilds#enable-prebuilt-workspaces).

> Next time you push code to your repository, the `prebuild` command will download all the dependencies and make the project startup faster.

Reviewer: Same as [this comment](https://github.com/gitpod-io/website/pull/2151/files/443b19ea019bfd7e375bb97e44a434f3ca676241#r891776458). We should update this to `init`

```suggestion
Next time you push code to your repository, the `init` command will download all the dependencies and make the project startup faster.
```

<!-- NEW PAGE --> 

## Getting Started / Quick Start

> <!-- TODO: @nancy --> 

Author: CC: @Nancy-Chauhan

4. Finally, start debugging your Java program!

You can also create the Java debug configuration file manually

To start debugging your Java application in Gitpod, please create a new directory called `.theia/`, and inside add a file called `launch.json`, finally add the following to it:

```json
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  "version": "0.2.0",
  "configurations": [
    {
      "type": "java",
      "name": "Debug (Launch) - Current File",
      "request": "launch",
      "mainClass": "${file}"
    }
  ]
}
```

Then, simply open the Java file you want to debug, open the Debug panel (in the left vertical toolbar, click the icon with the crossed-out-spider), and click the green "Run" button.

<br>

To see a basic repository with Java debugging enabled, please check out [gitpod-io/Gitpod-Java-Debugging](https://github.com/gitpod-io/Gitpod-Java-Debugging):

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/gitpod-io/Gitpod-Java-Debugging)

For more please see [VSCode's docs](https://code.visualstudio.com/docs/java/java-debugging)

> ### VSCode extensions

Reviewer 2: It would be useful to add a YAML snippet for copy & pasting the vscode extensions install, without having to dig the extension ids from OpenVSX

Gitpod comes equipped with the following VS Code extensions:

- [Language Support for Java(TM)](https://marketplace.visualstudio.com/items?itemName=redhat.java)
- [Debugger for Java](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-debug)
- [Java Dependency Viewer](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-dependency)

You can [install additional extensions](/docs/ides-and-editors/vscode-extensions) for your project if you want.

Most of the information you find in the [Java for VS Code](https://code.visualstudio.com/docs/languages/java) documentation applies to Gitpod as well.

> ## Vaadin in Gitpod

Reviewer 2: Minor - but for consistency when we introduce SDKMAN we have a one line to describe what it is. 

For Vadin we don't - I appreciate that for some pro Java user it might be obvious.

I would add a simple description of the tool, for consistency and for making docs accessible

Author: I'm going to suggest we leave out Vaadin for now, and bring back / introduce frameworks in a future iteration.

| [Vaadin starter](https://github.com/vaadin/skeleton-starter-flow-spring) | An empty starter for a new Vaadin project | [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/vaadin/skeleton-starter-flow-spring) |

</div>

### Configuring Gitpod for a Vaadin project

Start by downloading a [**Vaadin and Spring Boot** project starter](https://vaadin.com/start) if you don't have a Vaadin project from before.

Next, add a [`.gitpod.yml`](/docs/config-gitpod-file) file with the follwing content to the root of the project:

```YAML
tasks:
  - command: mvn spring-boot:run
ports:
  - port: 8080
    onOpen: open-preview
```

This will start the development server and open up the application in the preview browser window.

You are now ready to push your code to GitHub.

### Enable prebuilds for a faster startup

The first time you start a Vaadin application, it downloads both Maven and npm dependencies, which can take some time. You can speed up the GitPod startup by enabling [prebuild](/docs/prebuilds) for the repository.

Update your [`.gitpod.yml`](/docs/config-gitpod-file) file with the following content:

```YAML
tasks:

>   - prebuild: mvn install spring-boot:start spring-boot:stop -DskipTests

Reviewer 2: We should rewrite this using init instead. The key prebuild is deprecated

Author: Agree, fixed in the new docs. 

![image](https://user-images.githubusercontent.com/5528307/174260636-d716355a-87c9-49aa-aa5c-2c935f78dd4f.png)
